### ISSUE TITLE
ShareDialog using mode of shareSheet is not work on iOS over 11.x

### ISSUE THREAD

**Issue by fsseven** [facebook/facebook-ios-sdk#1938]
### Checklist before submitting a bug report

- [X] I've updated to the latest released version of the SDK
- [X] I've searched for existing [GitHub issues](https://github.com/facebook/facebook-ios-sdk/issues)
- [X] I've looked for existing answers on [Stack Overflow](https://facebook.stackoverflow.com), the [Facebook Developer Community Forum](https://developers.facebook.com/community/) and the [Facebook Developers Group](https://www.facebook.com/groups/fbdevelopers)
- [X] I've read the [Code of Conduct](https://github.com/facebook/facebook-ios-sdk/blob/main/CODE_OF_CONDUCT.md)
- [X] This issue is not security related and can safely be disclosed publicly on GitHub

### Xcode version

12.0

### Facebook iOS SDK version

12.1.0

### Dependency Manager

CocoaPods

### SDK Framework

Share

### Goals

I want to show ShareDialog using shareSheet mode to share contents on some devices of iOS 13

### Expected results

The ShareDialog should be present correctly

### Actual results

The ShareDialog cannot be presented using shareSheet mode, but it worked with other mode like native or web

### Steps to reproduce

I'm using the code below to show ShareDialog just like the code in demo:
```
let dialog = ShareDialog(viewController: vc, content: content, delegate: standard)
dialog.mode = .shareSheet
dialog.show()
```
However, this is not worked on some devices which makes me confused.
So I looked into the source code of facebook iOS sdk, and found that in facebook sdk **11.1.0** and below,
it used a workaround to fix the problem that `[SLComposeViewController isAvailableForServiceType:fbsdkdfl_SLServiceTypeFacebook()]` may return `NO` on iOS over 11.x
This workaround actually works as expected:
```
// FBShareKit/FBSDKShareDialog.m
- (BOOL)_canShowShareSheet
{
  if (![self.class.internalUtility isFacebookAppInstalled]) {
    return NO;
  }
 
  // iOS 11 returns NO for `isAvailableForServiceType` but it will still work
  NSOperatingSystemVersion iOS11Version = { .majorVersion = 11, .minorVersion = 0, .patchVersion = 0 };
  BOOL operatingSystemIsAdequate = [NSProcessInfo.processInfo isOperatingSystemAtLeastVersion:iOS11Version];
  BOOL composerIsAvailable = [self.class.socialComposeViewControllerFactory canMakeSocialComposeViewController]; // <-Look at here.
  return operatingSystemIsAdequate || composerIsAvailable;
}
```

But, start from Facebook SDK **11.2**, it abstracts the composeViewController and use `SocialComposeViewControllerFactory` to handle if compose controller can be made. 
```
// FBSDKSocialComposeViewControllerFactory.m
- (BOOL)canMakeSocialComposeViewController
{
  //  Look at here
  return [SLComposeViewController isAvailableForServiceType:FBSDKSocialComposeServiceType];
}
 
- (nullable id<FBSDKSocialComposeViewController>)makeSocialComposeViewController
{
  if (self.canMakeSocialComposeViewController) { 
    return [SLComposeViewController composeViewControllerForServiceType:FBSDKSocialComposeServiceType];
  } else {
    return nil;
  }
}
```
We can see that no workaround when make the composeViewController, which means on some iOS the `shareSheet` mode will not take effect.

**Comment by jawwad** [comment]
Hi @fsseven,

Can you explain a bit more about the issue?

In the Goals you mention:
> I want to show ShareDialog using shareSheet mode to share contents on some devices of iOS 13

However from the title, it seems like the shareSheet is not working at all. Could you explain a bit more about where you are encountering this issue and if it is just on certain versions of iOS? This would help us in troubleshooting the issue.

**Comment by arusu0629** [comment]
@jawwad 
Sorry for the side note. I had the same problem.
It was confirmed on iOS 15.0 and 15.1. (Using FBShareKit 12.1.0 from cocoapods)

My sample code
```swift
let dialog = ShareDialog(viewController: self.viewController, content: content, delegate: delegate)
dialog.mode = .shareSheet
print("$$$ before called show")
dialog.show()
print("$$$ after called show")
```

Here is the console log from that time
```
$$$ before called show
2021-11-16 11:59:48.010357+0900 develop_debug[5031:1583672] [core] isAvailableForServiceType: for com.apple.social.facebook returning NO
2021-11-16 11:59:48.010769+0900 develop_debug[5031:1583672] [core] isAvailableForServiceType: for com.apple.social.facebook returning NO
$$$ after called show
2021-11-16 11:59:48.011801+0900 develop_debug[5031:1585188] FBSDKLog: starting with Graph API v2.4, GET requests for /269762600539651/aem_conversion_configs should contain an explicit "fields" parameter
2021-11-16 11:59:48.012492+0900 develop_debug[5031:1585188] FBSDKLog: Starting with v13 of the SDK, a client token must be embedded in your client code before making Graph API calls. Visit https://developers.facebook.com/docs/ios/getting-started#step-3---configure-your-project to learn how to implement this change.
2021-11-16 11:59:48.220858+0900 develop_debug[5031:1585188] [] __nwlog_err_simulate_crash simulate crash already simulated "nw_hash_table_apply called with null table"
2021-11-16 11:59:48.221517+0900 develop_debug[5031:1585188] [] nw_hash_table_apply called with null table, dumping backtrace:
        [arm64] libnetcore-2736.12.1
    0   libnetwork.dylib                    0x000000018543d9c0 __nw_create_backtrace_string + 120
    1   libnetwork.dylib                    0x0000000184bb787c nw_hash_table_apply + 2808
    2   libnetwork.dylib                    0x000000018513f1e4 D810C689-1339-3FC6-96E8-196B9901E4DC + 6300132
    3   libnetwork.dylib                    0x000000018513ccc0 D810C689-1339-3FC6-96E8-196B9901E4DC + 6290624
    4   CFNetwork                           0x00000001848e8ff4 _CFURLStorageSessionCopyIdentifier + 20652
    5   libnetwork.dylib                    0x0000000184cdcda4 D810C689-1339-3FC6-96E8-196B9901E4DC + 1703332
    6   libdispatch.dylib                   0x00000001081b4a20 _dispatch_call_block_and_release + 32
    7   libdispatch.dylib                   0x00000001081b6700 _dispatch_client_callout + 20
    8   libdispatch.dylib                   0x00000001081c0e34 _dispatch_workloop_invoke + 2960
    9   libdispatch.dylib                   0x00000001081cbfa0 _dispatch_workloop_worker_thread + 916
    10  libsystem_pthread.dylib             0x00000001f3a311b0 _pthread_wqthread + 288
    11  libsystem_pthread.dylib             0x00000001f3a30f50 start_wqthread + 8
```
Sharesheet didn't show.

By the way, I updated the SDK from 11.0.1 to 12.1.0.
Did you change the behavior of `dialog.mode = .automatic` ? 

In 11.0.1, if the FB app was installed, the share sheet was displayed, otherwise the native screen(modal window) was displayed

But in 12.1.0, the native screen is always displayed regardless of whether the app is installed or not!

Please check this when you are free.
Thank you.

**Comment by fsseven** [comment]
@jawwad 

> Hi @fsseven,
> 
> Can you explain a bit more about the issue?
> 
> In the Goals you mention:
> 
> > I want to show ShareDialog using shareSheet mode to share contents on some devices of iOS 13
> 
> However from the title, it seems like the shareSheet is not working at all. Could you explain a bit more about where you are encountering this issue and if it is just on certain versions of iOS? This would help us in troubleshooting the issue.

@jawwad I'm sorry for the incorrect title. 
The problem actually occurs on iOS 12 or higher iOS versions.
For more detail, iPad Pro 11 inch with iOS 13.3.1 will meet the problem. 
Maybe a video will be more helpful to describe the problem?

**Comment by jawwad** [comment]
> Maybe a video will be more helpful to describe the problem?

A video would be amazing! Anything that will let us clearly see the issue and troubleshoot it faster will definitely help!

**Comment by samodom** [comment]
@fsseven Many thanks for reporting this issue.  We're pretty confident we have fixed the issue and our upcoming v12.2.0 release should include the resolution.  If you find that you still have an issue after adopting that version, please open a new issue and provide us any information that would help us reproduce the error.

**Comment by dianjakawidiawan50** [comment]
`<script data-ad-client="ca-pub-1091812641825939" async src="https://pagead2.googlesyndication.com/pagead/js/adsbygoogle.js"></script>`
